File: backend/app/schemas/brsr.py
"""BRSR export and wizard schemas."""

from __future__ import annotations

import uuid
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.services.reports.brsr_profile import AssuranceLevel, BrsrManualKpi, BrsrOrgProfileUpdate


class BrsrExportRequest(BaseModel):
    project_id: uuid.UUID | None = None
    reporting_year: int | None = Field(default=None, ge=2000, le=2100)
    format: Literal["json", "xlsx", "zip"] = Field(
        default="zip",
        description="json | xlsx | zip (JSON + Excel assurance pack)",
    )


class BrsrProfileOut(BaseModel):
    reporting_year: int | None = None
    listed_entity: bool = False
    cin: str | None = None
    stock_exchange: str | None = None
    assurance_level: AssuranceLevel = "none"
    boundary_notes: str | None = None
    manual_kpis: dict[str, BrsrManualKpi] = Field(default_factory=dict)
    wizard_completed_steps: list[str] = Field(default_factory=list)
    disclosure_complete: bool = False


class BrsrReadinessKpiOut(BaseModel):
    kpi_id: str
    name: str
    data_available: bool
    value_summary: str | None = None
    platform_source: str | None = None
    notes: str | None = None
    action: str


class BrsrReadinessOut(BaseModel):
    readiness_pct: int
    kpi_available_count: int
    kpi_total: int
    disclosure_complete: bool
    kpis: list[BrsrReadinessKpiOut]
    value_chain: dict[str, Any]
    blockers: list[str]
    export_ready: bool


class BrsrWizardStateOut(BaseModel):
    profile: BrsrProfileOut
    scope: str
    project_id: str | None = None
    reporting_year: int
    readiness: BrsrReadinessOut
    preview: dict[str, Any]


class BrsrProfilePatchIn(BrsrOrgProfileUpdate):
    pass

File: frontend/lib/brsr.ts
/** BRSR Principle 6 wizard API helpers and types. */

import { api } from "@/lib/api";

export type BrsrAssuranceLevel = "none" | "limited" | "reasonable";

export type BrsrManualKpi = {
  value_summary: string;
  source?: string | null;
};

export type BrsrProfile = {
  reporting_year: number | null;
  listed_entity: boolean;
  cin: string | null;
  stock_exchange: string | null;
  assurance_level: BrsrAssuranceLevel;
  boundary_notes: string | null;
  manual_kpis: Record<string, BrsrManualKpi>;
  wizard_completed_steps: string[];
  disclosure_complete: boolean;
};

export type BrsrReadinessKpi = {
  kpi_id: string;
  name: string;
  data_available: boolean;
  value_summary: string | null;
  platform_source: string | null;
  notes: string | null;
  action: string;
};

export type BrsrWizardState = {
  profile: BrsrProfile;
  scope: string;
  project_id: string | null;
  reporting_year: number;
  readiness: {
    readiness_pct: number;
    kpi_available_count: number;
    kpi_total: number;
    disclosure_complete: boolean;
    kpis: BrsrReadinessKpi[];
    value_chain: {
      project_count: number;
      missing_supplier_count: number;
      missing_supplier_projects: Array<{
        project_code: string;
        project_name: string;
        scheme_code?: string | null;
      }>;
    };
    blockers: string[];
    export_ready: boolean;
  };
  preview: {
    organization: { name: string };
    core_kpi_mapping: unknown[];
    value_chain_annex: unknown[];
    essential_indicator_ids: string[];
  };
};

export type BrsrWizardStep = "disclosure" | "scope" | "kpis" | "value_chain" | "export";

export const BRSR_WIZARD_STEPS: { id: BrsrWizardStep; label: string; hint: string }[] = [
  {
    id: "disclosure",
    label: "Org disclosure",
    hint: "CIN, reporting year, and assurance scope for listed entities.",
  },
  {
    id: "scope",
    label: "Portfolio scope",
    hint: "Export the full org portfolio or a single plantation project.",
  },
  {
    id: "kpis",
    label: "KPI readiness",
    hint: "Review P6.E1–E8 essential indicators and fill gaps.",
  },
  {
    id: "value_chain",
    label: "Value chain",
    hint: "Supplier references for P6.E8 nature-related value-chain evidence.",
  },
  {
    id: "export",
    label: "Review & export",
    hint: "Download the BRSR assurance pack for board or auditor review.",
  },
];

export const brsrApi = {
  readiness(projectId?: string) {
    return api
      .get<BrsrWizardState>("/v1/reports/brsr/readiness", {
        params: projectId ? { project_id: projectId } : undefined,
      })
      .then((r) => r.data);
  },
  updateProfile(payload: Partial<BrsrProfile>) {
    return api.patch<BrsrProfile>("/v1/reports/brsr/profile", payload).then((r) => r.data);
  },
};
